x-pack/plugins/apm/server/routes/source_maps/schedule_source_map_migration.ts

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import { FleetStartContract } from '@kbn/fleet-plugin/server';
+import { FleetArtifactsClient } from '@kbn/fleet-plugin/server/services';
+import { TaskManagerSetupContract } from '@kbn/task-manager-plugin/server';
+import { CoreStart, Logger } from '@kbn/core/server';
+import { getApmArtifactClient } from '../fleet/source_maps';
+import { bulkCreateApmSourceMaps } from './bulk_create_apm_source_maps';
+import { APM_SOURCE_MAP_INDEX } from '../settings/apm_indices/get_apm_indices';
+import { ApmSourceMap } from './create_apm_source_map_index_template';
+import { APMPluginStartDependencies } from '../../types';
+import { createApmSourceMapIndexTemplate } from './create_apm_source_map_index_template';
+
+const PER_PAGE = 10;
+const TASK_ID = 'apm-source-map-migration-task-id';
+const TASK_TYPE = 'apm-source-map-migration-task';
+
+export async function scheduleSourceMapMigration({
+  coreStartPromise,
+  pluginStartPromise,
+  fleetStartPromise,
+  taskManager,
+  logger,
+}: {
+  coreStartPromise: Promise<CoreStart>;
+  pluginStartPromise: Promise<APMPluginStartDependencies>;
+  fleetStartPromise?: Promise<FleetStartContract>;
+  taskManager?: TaskManagerSetupContract;
+  logger: Logger;
+}) {
+  if (!taskManager) {
+    return;
+  }
+
+  logger.debug(`Register task "${TASK_TYPE}"`);
+
+  taskManager.registerTaskDefinitions({
+    [TASK_TYPE]: {
+      title: 'Migrate fleet source map artifacts',
+      description:
+        'Migrates fleet source map artifacts to `.apm-source-map` index',
+      timeout: '24h',
```

Comment:
Do we anticipate that this task will take a really long time to run? Setting the timeout to `24h` means that task manager won't attempt to cancel the task for 24 hours, but it also means that the `retryAt` time is set to 24 hours after the task starts running the first time. If Kibana starts up and starts this task, then Kibana crashes while this task is running, the task will not retry again when Kibana restarts, instead it'll wait for 24 hours after the initial time. Is that acceptable?

Reply:
> Do we anticipate that this task will take a really long time to run?

It could if there are a lot of objects to go through. I can reduce it to 2 hours which is probably enough in 99% of cases but is better to cancel long running tasks than letting them run?